### PR TITLE
Allow sha256 state receipt

### PR DESCRIPTION
**Pull request by MatthewPhinney**

### CODE REVIEW

**Comment by clark800** [comment]
LGTM

**Comment by MatthewPhinney** [comment]
Responses from the `/transfers/:id/state` endpoint failed to validate against the `TransferStateReceipt` JSON schema because the schema did not allow type `sha256`, and required a few fields that were specific to `ed25519` (such as `public_key`). This pull request modifies the schema to allow for `sha256` receipts. sha-256 example:
```
{ type: 'sha256',
  message: 
   { id: 'http://localhost/transfers/3a2a1d9e-8640-4d2d-b06c-84f2cd613204',
     state: 'proposed',
     token: '9pHqGD4ySNg7O2hJvXy+XNp3mQFwm9lQIpayE7Id45bajfTZEKMapRL7Kotfu/plBfLn6KJL0cs5B31rR3CzAg==' },
  signer: 'http://localhost',
  digest: 'LSbOft0NAHoXrMtD9IWpAc3UQ/GR+FvIv9F1xMB/rVY=',
  condition_state: 'executed',
  condition_digest: 'x2+SVBZ1IHCmjgnicbxfTEH7Cr6QS+nebQIdC8osFpc=' }
```
ed25519 example:
```
{ type: 'ed25519-sha512',
  message: 
   { id: 'http://localhost/transfers/9e97a403-f604-44de-9223-4ec36aa466d9',
     state: 'executed' },
  signer: 'http://localhost',
  public_key: '9PAqTUEptSeQCOp/0FQTm3rkFnUFaYEUEwCcyyySQP0=',
  signature: '2tQy/3e8MH0UIpPXp0hWjYYUepqfaQLcNNUm0vu5eiArOVOMuzTmG2+1Cg8siD+jlF3uFxp3rKIxESY8NYpdDA==' }
```
We need to allow both types.  I also removed the `SignedMessageTemplate` and the verify methods in `utils` because they did not appear to be used anywhere. I ran this by Stefan, but I'd like @emschwartz  to sign off on this before it's merged.

File: schemas/TransferStateReceipt.json [hunk 1]
```diff
@@ -24,7 +46,19 @@
           }]
         }
       }
+    },
+    "condition_state": {
+      "description": "Transfer state",
```
Comment (MatthewPhinney) [inline schemas/TransferStateReceipt.json]: I'm actually not sure what `condition_state` is. Do we have a list of values for this field?

Comment (sentientwaffle) [inline schemas/TransferStateReceipt.json]: `condition_state` is a Transfer state, with an additional state `"nonexistent"`.
`"nonexistent"` is for when you fetch the state of a transfer that doesn't exist -- it is needed for atomic mode to get the receipt condition).

**Comment by MatthewPhinney** [comment]
Stefan and I discussed and decided to remove the unused verify methods. I think we use `five-bells-condition` for the verify functionality. Are those methods still needed? Ideally I'd like to have TransferStateReceipt inherit from two schemas, one for sha256, and one for ed25519, since each schema has different required fields, but from my <a href="http://spacetelescope.github.io/understanding-json-schema/reference/combining.html#allof">reading</a> and testing, it's not possible to inherit in the object oriented sense (trying to do so led to invalid JSON validating just fine). I'll try separating into two different files and using the `oneOf` syntax. Is the `SignedMessageTemplate` used for anything aside from the `verify` methods?

**Comment by emschwartz** [comment]
The `oneOf` solution sounds good to me. I would make a `SignedMessageTemplate` that includes a `oneOf` either sha256 or ed25519. You can look at https://github.com/interledger/five-bells-shared/blob/master/schemas/Transfer.json and https://github.com/interledger/five-bells-shared/blob/master/schemas/TransferTemplate.json as an example of how to extend a schema.

The reason it's important to separate the `SignedMessageTemplate` from the `TransferStateReceipt` is that in the future the receipt won't always be a `TransferStateReceipt`, but rather just a signed message from the recipient. Using the `TransferStateReceipt` as the execution condition is kind of a hack that we're using for now but will probably move away from in the future.

**Comment by MatthewPhinney** [comment]
I think it's a bit cleaner now. I created schemas for sha256 and ed25519, and then added the common field that both share (`message`) to SignedMessageTemplate.

**Comment by sentientwaffle** [comment]
This still won't validate `/state` -- the ledger returns the `signer` property for both types of receipts.

**Comment by MatthewPhinney** [comment]
@sentientwaffle I added a commit to include `signer` in `sha256` and to disallow additional properties. It actually validated before, since additional properties were allowed. In order to set `additionalProperties` to `false`, I had to include the `message` key in each schema (I fully specify this field in `SignedMessageTemplate`).

**Comment by sentientwaffle** [comment]
@MatthewPhinney `message` has a `token` field for sha256.

**Comment by MatthewPhinney** [comment]
Where is this? The schema in this PR validates against the ledger code (it could be incorrect, but it seems to work). I see a `token` field in the <a href="https://github.com/interledger/five-bells-ledger/blob/master/src/models/transfers.js#L81">"hash preimage"</a>, but I don't see where that is returned in the endpoint.

**Comment by sentientwaffle** [comment]
https://github.com/interledger/five-bells-ledger/blob/master/src/models/transfers.js#L64

**Comment by MatthewPhinney** [comment]
Nvm, I see it ...

**Comment by MatthewPhinney** [comment]
@sentientwaffle updated.

**Comment by sentientwaffle** [comment]
LGTM

**Pull request merged**